app/app.go

```diff
@@ -0,0 +1,169 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"runtime"
+
+	"x-gwi/app/auth"
+	"x-gwi/app/instance"
+	"x-gwi/app/logs"
+	"x-gwi/app/pki"
+	"x-gwi/app/server"
+	"x-gwi/app/storage"
+)
+
+type App struct {
+	config  *Config
+	inst    *instance.Instance
+	pki     *pki.PKI
+	auth    *auth.Auth
+	storage *storage.AppStorage
+	server  *server.Server
+
+	cancel context.CancelFunc
+	chErr  chan error
+	chMsg  chan string
+}
+
+func Main() {
+	if err := runApp(context.Background(), newConfig()); err != nil {
+		logs.Error().Err(err).Msg("app exit ERROR")
+		os.Exit(1)
+	}
+
+	logs.Info().Msg("app exit OK")
+	os.Exit(0)
+}
+
+func runApp(ctx context.Context, config *Config) error {
+	app, ctx, err := initApp(ctx, config)
+	if err != nil {
+		return err
+	}
+
+	app.server.Serve(ctx)
+	runtime.Gosched()
```

Comment:
Why? you are already calling gosched inside server.Serve()
Also, why do you call gosched after starting new goroutines?

Reply:
It is to control order of execution.  Here it is double check any go func in  inside server.Serve() will be scheduled before next line in app (48+). It can happen that  modifying quickly deep code I switch on/off go func so leaving runtime.Gosched() in main routine can help in these cases. It doesn't block if there is nothing to do and it helps to be more ordered for go func in any

runtime.Gosched() allows to process a go subroutine body before next line in 'parent' routine;
go scheduler behaves in way that it first process code in main routine and any go func is postponed until main riches end or sleep or waiting for chan input. In past I used to think go func behaves unordered but this finding shows they are processed in reverse order like defer

It is very useful for initialising waiting group where wg.Add(1) can be added inside go funct /not before (for example flexible number of started routines). Then adding line runtime.Gosched() before wg.Wait() ensures wg.Add(1) in subroutine is processed before wg.Wait(), (otherwise wg.Wait() passes because scheduler didn't process  wg.Add(1) in subroutine)

```sh
for i := 0; i < N; i++ {
go func() {
	wg.Add(1)
	wg.Done()
}()
}
runtime.Gosched()
wg.Wait()

```

Doe it answer your question?

Reply:
Yes, it does. However this is wrong.

There are 2 issues here.

1) you *must not ever* call `wg.Add` from inside the goroutine you are spawning. Because goroutine scheduling is not deterministic and you are NEVER guaranteed that the goroutine you spawned (even when you call gosched) gets executed before you call `wg.Wait`.

Firstly it specifically says so in the docs: https://pkg.go.dev/sync#WaitGroup.Add : ` Add should execute before the statement creating the goroutine or other event to be waited for`

And secondly, this creates a bug that is very very difficult to even find, because in some cases one of your goroutines never gets spawned, but the code looks ok.

2) the runtime.Gosched() should not be used in this type of code, it is used here as a workaround to wrong usage of wg.Add. The main usecases for gosched is for I/O where you can yield your code to allow different goroutines to run. In standard library, the usecases that can be found are: right after calling syscall, signal handling, mutex operations, GC triggers (there are more, feel free to check for yourself: https://cs.opensource.google/search?q=gosched&sq=&ss=go%2Fgo)

Reply:
I agree it is potentially not clear and we should be clear not clever with GO :)

From other hand it doesn't break the docs because it says about execution of Add before the statement (not about placement of statement Add) and we can achieve it in more ways: first clear is to put statement Add before statement go doing add; second is to strictly ensure execution of Add before statement waiting for go statement(s) doing add for example by gosched. And once we have more than two ways we take control on execution of our code. Plus we can suggest an update of documentation ;)

As I had checked the order of execution of go subroutine(s) it is the same as order of execution of defer(s). In first iteration of scheduler they are only initialised with potential arguments. The difference is in triggers where defer is triggered by end of function body or by return while go subroutines are triggered by the scheduler sequentially assigning execution time for main routine and separate subroutines (if main routine not end before).

Example of Add is simple, easy to replace and test each approach, so nice to discuss and learn

Other place where it is used is starting the server where sequence os steps using subroutines is required. It is easy with gosched. Other ways are unequivocal to answer if grpc is listening on net port. Checking too quickly by client can return err. Adding sleep time has more disadvantages then gosched. So with my current knowledge I keep gosched as recommended there for simplicity, control and performance.

Reply:
> As I had checked the order of execution of go subroutine(s) it is the same as order of execution of defer(s).

Can you please provide your source for this? 

Please note this is irrelevant to the application, I'm just curious, because I don't think this is correct. I looked at this: https://go.dev/src/runtime/proc.go?s=8144:8158 and https://docs.google.com/document/d/1TTj4T2JO42uD5ID9e89oa0sLKhJYD0Y_kqxDv3I3XMw/edit#heading=h.1amka2n5suqo where this one part about scheduling would lead to non-deterministic behaviour depending on each processors list of runnable goroutines:
```
When a new G is created or an existing G becomes runnable, it is pushed onto a list of runnable goroutines of current P. When P finishes executing G, it first tries to pop a G from own list of runnable goroutines; if the list is empty, P chooses a random victim (another P) and tries to steal a half of runnable goroutines from it.
```

Specifically this part:
```
if the list is empty, P chooses a random victim (another P) and tries to steal a half of runnable goroutines from it.
```

So depending on the total number of goroutines in each of processors list of goroutines, it can lead to condition where they would be running in very different order from the code, even after running gosched.

Defers work differently from this, and are always deterministic.

Reply:
I have investigated behaviour playing with code like one below:
add func, trigger, remove=comment, change order ... (below is some state after adding and checking some stats)

The ordered behaviour appears after adding runtime.Gosched() which reverse orders closest go func (like defer)
If runtime.Gosched() is not used go func runs in unpredictable order (not like defer) = like you described above and like I used to think

imo we can switch on order, for at least closest subroutines, by placing runtime.Gosched() otherwise subroutines runs in kind of random

```go
package main

import (
	"fmt"
	"runtime"
	"time"
)

var t time.Time

func say5(s string) {
	// go func() {
	// 	wg.Add(1)
	// 	wg.Done()
	// }()
	// runtime.Gosched()
	// wg.Wait()
	for i := 0; i < 5; i++ {
		t1 := time.Since(t)
		runtime.Gosched()
		t2 := time.Since(t)
		fmt.Println(i+1, s, runtime.NumCPU(), runtime.NumGoroutine(), t1, t2)
	}
}

func main() {
	fmt.Println(runtime.Version())
	t = time.Now()
	fmt.Println("start", runtime.NumCPU(), runtime.NumGoroutine(), time.Since(t))
	go say5("world-sub-1") // create a new goroutine
	go say5("world-sub-2") // create a new goroutine
	fmt.Println("start", runtime.NumCPU(), runtime.NumGoroutine(), time.Since(t))
	// time.Sleep(time.Millisecond)
	say5("hello-main-") // current goroutine
	for i := 0; i < 6; i++ {
		t1 := time.Since(t)
		runtime.Gosched()
		t2 := time.Since(t)
		fmt.Println(i+1, "main-sched-", runtime.NumCPU(), runtime.NumGoroutine(), t1, t2)
	}
	fmt.Println("end", runtime.NumCPU(), runtime.NumGoroutine(), time.Since(t))
	// time.Sleep(time.Millisecond)
	fmt.Println("end", runtime.NumCPU(), runtime.NumGoroutine(), time.Since(t))
}
```

Reply:
I have fixed wg.Add to be more clear and GO way, although discussion was very interesting :)

Also runtime.Gosched() is removed.
In part starting server there is grpc client trying connection before rest gw. It is set to use option grpc.WithBlock(). Usually this option is not required but here is added by intention with purpose to ensure working connection before starting rest gw